sg14-error-codes.html

```diff
@@ -620,16 +658,26 @@
         </li>
     </ul>
     Arthur O'Dwyer's (admittedly uninformed) opinion is that these sound like antipatterns that could reasonably be avoided.
-    Niall Douglas's opinion seems to be either that these are valid patterns, or else that they turn out to be impossible to
-    avoid in practice. Charley Bay has also seen these failure modes in practice.
+    Niall Douglas's opinion is that these are patterns in use in big consumers e.g. Python loads its C/C++ modules with
+    <code>RTLD_LOCAL</code>, and indeed has no choice but to do so. Charley Bay has also seen these failure modes in practice.
```

Comment:
Confirmed that Python loads modules with `RTLD_LOCAL`.  So the possible failure mode here is that you have a Python app that imports two C++ modules, and both C++ modules have their own copies of the typeinfo for `Foo`; and then the Python app calls into the first C++ module, gets a `Foo` object, passes that object in to the second C++ module, and the second C++ module doesn't believe it's a `Foo` because its typeid doesn't compare address-equal. Is that right?
(Or repeat the same setup but with `s/typeid/error_category/` and `s/Foo/error_code/`.)

However, this wouldn't be a pure C++-language program, right? it'd be a Python program that uses C++ modules? and if we rewrote the Python app *in* C++ so that we-the-programmer (not Python) controlled the parameters to `dlopen`, then we could easily fix the problem by just using `RTLD_GLOBAL`?  Now, rewriting the Python app in C++ is not likely to be a reasonable course of action; but by picking a Python app as our example, are we asking the *C++* committee to care deeply about the correctness of *Python* apps? I mean, this might be the rare case where someone is justified in saying, "Just write it in C++ and all your problems will go away." ;)

Reply:
Lots and lots of stuff is forced to load shared objects with RTLD_LOCAL because ELF is so badly broken. It's hardly just Python.

We really, seriously, need a new standardised binary loader in C++, with a new universal binary format which can also still be packaged as ELF/PE/MachO/whatever.